### PR TITLE
Build with Earthly

### PR DESCRIPTION
**Pull request by darabos**
For #295. I'm just learning this and the setup is not perfect yet. But it's working! To try:
- [Install Earthly](https://earthly.dev/get-earthly)
- Run `earthly +run`.
- Go to http://localhost:2200/.

It looks 100% great for builds! As the next step I'll try running tests inside and set up caching. The setup I'm going for:
- Add `SAVE IMAGE` strategically in places worth caching.
- Set it up with Google Artifact Registry. Use secrets in GitHub Actions.
- Add a new GitHub Action that pushes the images to cache when a PR is merged.

With that my hope is that GitHub Actions on PRs will be super fast because they only run the tests that are impacted.

### CODE REVIEW

**Comment by darabos** [comment]
The frontend test fails:

```
      +frontend-test | [12:17:03] I/update - chromedriver: setting permissions to 0755 for /home/mambauser/lk/web/node_modules/webdriver-manager/selenium/chromedriver_2.35
      +frontend-test | [12:17:03] I/downloader - curl -o/home/mambauser/lk/web/node_modules/webdriver-manager/selenium/selenium-server-standalone-3.141.59.jar https://selenium-release.storage.googleapis.com/3.141/selenium-server-standalone-3.141.59.jar
      +frontend-test | [12:17:09] I/downloader - curl -o/home/mambauser/lk/web/node_modules/webdriver-manager/selenium/geckodriver-v0.32.0.tar.gz https://github.com/mozilla/geckodriver/releases/download/v0.32.0/geckodriver-v0.32.0-linux64.tar.gz
      +frontend-test | [12:17:09] I/update - geckodriver: unzipping geckodriver-v0.32.0.tar.gz
      +frontend-test | [12:17:09] I/update - geckodriver: setting permissions to 0755 for /home/mambauser/lk/web/node_modules/webdriver-manager/selenium/geckodriver-v0.32.0
      +frontend-test | [12:17:09] Finished 'webdriver-update' after 6.6 s
      +frontend-test | [12:17:09] Starting 'test'...
      +frontend-test | [12:17:09] W/driverProviders - Using driver provider directConnect, but also found extra driver provider parameter(s): seleniumServerJar
      +frontend-test | [12:17:09] I/launcher - Running 1 instances of WebDriver
      +frontend-test | [12:17:09] I/direct - Using ChromeDriver directly...
      +frontend-test | [12:17:10] E/launcher - Server terminated early with status 127
      +frontend-test | [12:17:10] E/launcher - Error: Server terminated early with status 127
      +frontend-test |     at /home/mambauser/lk/web/node_modules/selenium-webdriver/remote/index.js:252:52
      +frontend-test |     at processTicksAndRejections (node:internal/process/task_queues:95:5)
      +frontend-test | From: Task: WebDriver.createSession()
      +frontend-test |     at Function.createSession (/home/mambauser/lk/web/node_modules/selenium-webdriver/lib/webdriver.js:769:24)
      +frontend-test |     at Function.createSession (/home/mambauser/lk/web/node_modules/selenium-webdriver/chrome.js:761:15)
      +frontend-test |     at Direct.getNewDriver (/home/mambauser/lk/web/node_modules/protractor/built/driverProviders/direct.js:77:33)
      +frontend-test |     at Runner.createBrowser (/home/mambauser/lk/web/node_modules/protractor/built/runner.js:195:43)
      +frontend-test |     at /home/mambauser/lk/web/node_modules/protractor/built/runner.js:339:29
      +frontend-test |     at _fulfilled (/home/mambauser/lk/web/node_modules/q/q.js:834:54)
      +frontend-test |     at /home/mambauser/lk/web/node_modules/q/q.js:863:30
      +frontend-test |     at Promise.promise.promiseDispatch (/home/mambauser/lk/web/node_modules/q/q.js:796:13)
      +frontend-test |     at /home/mambauser/lk/web/node_modules/q/q.js:556:49
      +frontend-test |     at runSingle (/home/mambauser/lk/web/node_modules/q/q.js:137:13)
      +frontend-test | [12:17:10] E/launcher - Process exited with error code 199
```

It's working on GitHub Actions with the same super old Chromedriver version. It's not clear why it fails in Earthly. But I think it's about time we fix up the tests anyway. Chromedriver 2.35 is compatible up to Chrome 64, released in 2018. I'm not sure what's responsible for installing this ancient version of Chrome, but it may just be impossible in Ubuntu 22.04 (Jammy), which I used as the base image here.

It's another huge task, but I think rewriting the tests in https://playwright.dev/ would be a good investment.

**Comment by darabos** [comment]
The cache is working!!!
image

It's not 100% working. It got the base image from cache, which is the important bit with the Conda environment. After that it built everything and ran the tests.

I think it will get better. I think the cause of the cache miss is that my local repo is full of garbage so some of the larger `COPY` commands pick up more for me than on GitHub Actions. This will not be an issue when I merge this and trigger the new GitHub Action that updates the cache!

But even if it doesn't get better, 13 minutes is fine. 👍

**Comment by darabos** [comment]
Alright, let's get this and #292 merged! If there are any late review comments, let me know, @lacca0.

File: test/com/lynxanalytics/biggraph/frontend_operations/GraphMetricsTest.scala [hunk 1]
```diff
@@ -15,7 +15,7 @@ class GraphMetricsTest extends OperationsTestBase {
   }
 
   // All 2 plot tests use this box
-  val plotBox = box("Create example graph")
+  def plotBox = box("Create example graph")
```
Comment (lacca0) [inline test/com/lynxanalytics/biggraph/frontend_operations/GraphMetricsTest.scala:18]: What is this change for?

**Review by lacca0 — APPROVED** [review]

File: test/com/lynxanalytics/biggraph/frontend_operations/GraphMetricsTest.scala (hunk 1 above)
Comment (darabos) [reply test/com/lynxanalytics/biggraph/frontend_operations/GraphMetricsTest.scala:18]: > What is this change for?

With `val` the right-hand side is executed when the test definitions are loaded. With `def` it's only executed when the tests are running.

`scalatest` loads the test definitions for all tests to see which tests match the requested tags. So this line was triggered no matter which tags we are testing. But when running the RequiresDocker tests (`+backend-test-docker`) I didn't copy in the `built-ins` directory, so this line was causing a failure.

There are two solutions: a) copy in `built-ins` even if we are not running tests that use it, or b) don't touch `built-ins` when no tests running need it. I went with "b)".

**Pull request merged**